File: zig/string_reverse.zig
const std = @import("std");
const print = std.debug.print;

fn reverseString(allocator: std.mem.Allocator, str: []const u8) ![]u8 {
    const result = try allocator.alloc(u8, str.len);
    var start: usize = 0;
    var end: usize = str.len - 1;
    
    @memcpy(result, str);
    
    while (start < end) {
        const temp = result[start];
        result[start] = result[end];
        result[end] = temp;
        start += 1;
        end -= 1;
    }
    
    return result;
}

fn generateString(allocator: std.mem.Allocator, length: usize) ![]u8 {
    const str = try allocator.alloc(u8, length);
    for (0..length) |i| {
        str[i] = @as(u8, 'A') + @as(u8, @intCast(i % 26));
    }
    return str;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();
    
    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);
    
    if (args.len < 2) {
        print("Usage: string_reverse <length>\n", .{});
        return;
    }
    
    const length = try std.fmt.parseInt(usize, args[1], 10);
    const str = try generateString(allocator, length);
    defer allocator.free(str);
    
    const start = std.time.nanoTimestamp();
    const reversed = try reverseString(allocator, str);
    defer allocator.free(reversed);
    const end = std.time.nanoTimestamp();
    
    // Calculate checksum to verify correctness
    var checksum: i64 = 0;
    for (reversed) |c| {
        checksum += c;
    }
    
    const time_ms = @divFloor(end - start, 1_000_000);
    
    print("Zig: string_reverse({d}) = {d}\n", .{ length, checksum });
    print("Time: {d}ms\n", .{time_ms});
}

File: c/string_reverse.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>

void reverse_string(char* str, int len) {
    int start = 0;
    int end = len - 1;
    
    while (start < end) {
        char temp = str[start];
        str[start] = str[end];
        str[end] = temp;
        start++;
        end--;
    }
}

char* generate_string(int len) {
    char* str = malloc((len + 1) * sizeof(char));
    if (!str) return NULL;
    
    for (int i = 0; i < len; i++) {
        str[i] = 'A' + (i % 26);
    }
    str[len] = '\0';
    return str;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        printf("Usage: string_reverse <length>\n");
        return 1;
    }
    
    int len = atoi(argv[1]);
    char* str = generate_string(len);
    if (!str) {
        printf("Memory allocation failed\n");
        return 1;
    }
    
    clock_t start = clock();
    reverse_string(str, len);
    clock_t end = clock();
    
    double time_ms = ((double)(end - start) / CLOCKS_PER_SEC) * 1000;
    
    // Calculate checksum to verify correctness
    long checksum = 0;
    for (int i = 0; i < len; i++) {
        checksum += str[i];
    }
    
    printf("C: string_reverse(%d) = %ld\n", len, checksum);
    printf("Time: %.0fms\n", time_ms);
    
    free(str);
    return 0;
}